Bugzilla/Report/SecurityRisk.pm

```diff
@@ -0,0 +1,318 @@
+                Dict [
+                    open   => ArrayRef [Int],
+                    closed => ArrayRef [Int],
+                    median_age_open => Num
+                ]
+            ],
+            bugs_by_sec_keyword => HashRef [
+                Dict [
+                    open   => ArrayRef [Int],
+                    closed => ArrayRef [Int],
+                    median_age_open => Num
+                ]
+            ],
+        ],
+    ],
+);
+
+sub _build_initial_bug_ids {
+    # TODO: Handle changes in product (e.g. gravyarding) by searching the events table
+    # for changes to the 'product' field where one of $self->products is found in
+    # the 'removed' field, add the related bug id to the list of initial bugs.
+    my ($self) = @_;
+    my $dbh = Bugzilla->dbh;
+    my $products     = join ', ', map { $dbh->quote($_) } @{ $self->products };
+    my $sec_keywords = join ', ', map { $dbh->quote($_) } @{ $self->sec_keywords };
+    my $query        = qq{
+        SELECT
+            bug_id
+        FROM
+            bugs AS bug
+            JOIN products AS product ON bug.product_id = product.id
+            JOIN components AS component ON bug.component_id = component.id
+            JOIN keywords USING (bug_id)
+            JOIN keyworddefs AS keyword ON keyword.id = keywords.keywordid
+         WHERE
+            keyword.name IN ($sec_keywords)
+            AND product.name IN ($products)
+    };
+    return Bugzilla->dbh->selectcol_arrayref($query);
+}
+
+sub _build_initial_bugs {
+    my ($self)    = @_;
+    my $bugs      = {};
+    my $bugs_list = Bugzilla::Bug->new_from_list( $self->initial_bug_ids );
+    for my $bug (@$bugs_list) {
+        my $cf_last_resolved = $bug->cf_last_resolved;
+        $bugs->{ $bug->id } = {
+            id        => $bug->id,
+            product   => $bug->product,
+            sec_level => (
+                # Select the first keyword matching one of the target keywords
+                # (of which there _should_ only be one found anyway).
+                first {
+                    my $x = $_;
+                    grep { lc($_) eq lc( $x->name ) } @{ $self->sec_keywords }
+                }
+                @{ $bug->keyword_objects }
+            )->name,
+            is_open    => $self->check_open_state->( $bug->status->name ),
+            created_at => datetime_from( $bug->creation_ts ),
+        };
+    }
+    return $bugs;
+}
+
+sub _build_events {
+    my ($self) = @_;
+    my $bug_ids    = join ', ', @{ $self->initial_bug_ids };
+    my $start_date = $self->start_date->ymd('-');
+    my $query      = qq{
+        SELECT
+            bug_id,
+            bug_when,
+            field.name AS field_name,
+            CONCAT(removed) AS removed,
+            CONCAT(added) AS added
+        FROM
+            bugs_activity
+            JOIN fielddefs AS field ON fieldid = field.id
+            JOIN bugs AS bug USING (bug_id)
+        WHERE
+            bug_id IN ($bug_ids)
```

Comment:
This crashes when $bug_ids is open which happened in my test environment. It worked fine after I created the sec-critical and sec-high keywords and assigned some bugs to them. This will not happen in prod probably ever so nothing to worry with. Otherwise we could return empty list [] of $self->initial_bug_ids is empty.

Reply:
Fixed, let me know if the `|| []` is the proper way to fix this :)

Reply:
Misread this problem, fixed it appropriately now!